### ISSUE TITLE
Floating point values should not go in advance widths

### ISSUE THREAD

**Issue by ctrlcctrlv** [rsms/inter#508]
```fish
#!/bin/bash
ADVANCES=$(rg 'advance width'|awk 'BEGIN {FS="  "} {print $2}')
NOTREALLYFLOATS=$((rg '\.0\b'|wc -l) <<< "$ADVANCES")
FLOATS=$((rg '\.\d+'|wc -l) <<< "$ADVANCES")
INTS=$((rg -v '\.'|wc -l) <<< "$ADVANCES")
printf "Total: $((INTS+FLOATS)) (sanity: $(wc -l <<< "$ADVANCES"))\nFloats: $((FLOATS-NOTREALLYFLOATS))\nInts: $((INTS+NOTREALLYFLOATS))\n"
```
```
Total: 85593 (sanity: 85593)
Floats: 40662
Ints: 44931
```

And if we narrow it down, these are your fractional components:
```fish
(rg -o '\.\d{1,12}'|sort|uniq -c) <<< "$ADVANCES"
```
```
  16470 .0
    207 .000000000000
    838 .199999999999
   6101 .2
    252 .200000000000
   5786 .333333333333
   1099 .399999999999
   5771 .4
    319 .400000000000
      2 .400000000001
    276 .599999999999
   5785 .6
   1130 .600000000000
   5766 .666666666666
     20 .666666666667
    244 .799999999999
   6062 .8
    885 .800000000000
    119 .999999999999
```

So really the most popular fractional component is pointless. `hmtx` can't hold these.

_Cf._ https://github.com/MFEK/glifparser.rlib/issues/54.

**Comment by rsms** [comment]
How do you run this script?
Could you add a bit more information about why this matters and what it is? (Is "advance widths" about glyphs or something else?)

**Comment by ctrlcctrlv** [comment]
Ah, sorry, I think this actually ought to be dealt with at googlefonts/glyphsLib#821. It's fundamentally an issue in that.

**Comment by rsms** [comment]
Ah. Thanks :-)

I think I figured out what it is: when I install ripgrep and run the script in the directory with my UFOs, I can repro your output.

Do you think it's worth postprocessing my UFOs to round these advance widths? (until it's been addressed in glyphsLib)

**Comment by ctrlcctrlv** [comment]
I do think so, yeah, if not _pre-process_ in Glyphsapp (I wouldn't know how sorry). The issue is really that the UFO spec makes a promise (advance width can be int or float) that the binary format (OpenType, TrueType) cannot satisfy. So, originally, [MFEK/glifparser.rlib](https://github.com/MFEK/glifparser.rlib) treated this as a hard error. But I downgraded that to a warning unless you specifically ask for any possible error in OT export to be flagged (OT strict output).

So all of these decimals will get dropped.

**Comment by rsms** [comment]
FWIW, the fractional advance widths are only produced when generating instances with `fontmake -i`, the master UFOs produced by fontmake from the glyphs file does not have fractional advance widths.

Running your script on the 12 master UFOs generated by fontmake:

```
Total: 28459 (sanity:    28459)
Floats: 0
Ints: 28459
```

Running your script on the 24 instance UFOs generated by fontmake:

```
Total: 57242 (sanity:    57242)
Floats: 40516
Ints: 16726
```

**Comment by rsms** [comment]
With [postprocessing](https://github.com/rsms/inter/commit/adf3654b33feafaed1741f8e27c3837c6e12715c#diff-236734d1413709a6d46d53e8adb03c762a75c73a5d66bd931a17cbdfaaf799f0R22):

```
Total: 85688 (sanity:    85688)
Floats: 0
Ints: 85688
```

**Comment by rsms** [comment]
To anyone coming across this in the future: There's a bug in defcon that prevents updating values where int(v)==int(v), so I ended up with a workaround that sets the value twice: https://github.com/rsms/inter/commit/9d9bfc3eecdd7ee17fe36827c400c8dcc0eb4cd5

**Comment by rsms** [comment]
Included in [v4.0-beta6](https://github.com/rsms/inter/releases/tag/v4.0-beta6)
